oauth2client/flask_util.py

```diff
@@ -0,0 +1,426 @@
+# Copyright 2015 Google Inc.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utilities for the Flask web framework
+
+Provides a Flask extension that makes using OAuth2 web server flow easier.
+The extension includes views that handle the entire auth flow and a @required
+decorator to automatically ensure that user credentials are available.
+
+To configure::
+
+    app = Flask(__name__)
+
+    app.config['SECRET_KEY'] = 'your-secret-key'
+
+    app.config['OAUTH2_CLIENT_SECRETS_JSON'] = 'client_secrets.json'
+    # or, specify the client id and secret separately
+    app.config['OAUTH2_CLIENT_ID'] = 'your-client-id'
+    app.config['OAUTH2_CLIENT_SECRET'] = 'your-client-secret'
+
+    oauth2 = OAuth2(app)
+
+
+To use::
+
+    # Note that app.route should be the outermost decorator.
+    @app.route('/needs_credentials')
+    @oauth2.required
+    def example():
+        # http is authorized with the user's credentials and can be used
+        # to make http calls.
+        http = oauth2.http()
+
+        # Or, you can access the credentials directly
+        credentials = oauth2.credentials
+
+
+    @app.route('/info')
+    @oauth2.required
+    def info():
+        return "Hello, {}".format(oauth2.email)
+
+    @app.route('/optional')
+    def optional():
+        if oauth2.has_credentials():
+            return 'Credentials found!'
+        else:
+            return 'No credentials!'
+
+"""
+
+__author__ = 'jonwayne@google.com (Jon Wayne Parrott)'
+
+import hashlib
+import json
+import os
+from functools import wraps
+
+import six.moves.http_client as httplib
+import httplib2
+
+try:
+    from flask import Blueprint
+    from flask import _app_ctx_stack
+    from flask import current_app
+    from flask import redirect
+    from flask import request
+    from flask import session
+    from flask import url_for
+except ImportError:
+    raise ImportError('The flask utilities require flask 0.9 or newer.')
+
+from oauth2client.client import FlowExchangeError
+from oauth2client.client import OAuth2Credentials
+from oauth2client.client import OAuth2WebServerFlow
+from oauth2client.client import Storage
+from oauth2client import clientsecrets
+
+
+DEFAULT_SCOPES = ('email',)
+
+
+class OAuth2(object):
```

Comment:
Naming this just OAuth2 on its own is a bit general for a framework-specific and user-auth-specific decorator. People might assume this is a more general purpose than it is. I would lean towards naming this OAuth2Decatorator just like the WebAbb2 one. Alternatively, something more scenario oriented like UserAuthDecorator. The duplicate naming should not be a problem because you don't use more than one framework at a time.